File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/mail_compose/MailComposeDtos.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.data.api.models.mail_compose

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

// GET /api/mailbox/compose/recipients

@JsonClass(generateAdapter = true)
data class MailComposeRecipientsResponse(
    val recipients: List<MailRecipientDto> = emptyList(),
)

@JsonClass(generateAdapter = true)
data class MailRecipientDto(
    val userId: String,
    val name: String? = null,
    val username: String? = null,
    val homeId: String? = null,
    val homeAddress: String? = null,
    val isVerified: Boolean? = null,
    val homeMediaUrl: String? = null,
    val isOnPantopus: Boolean? = null,
)

// GET /api/mailbox/compose/home-context/:homeId

@JsonClass(generateAdapter = true)
data class MailHomeContextResponse(
    val homeId: String? = null,
    val addressDisplay: String? = null,
    val memberCount: Int? = null,
    val homeMediaUrl: String? = null,
    val privateDeliveryAvailable: Boolean? = null,
    val members: List<MailHomeMemberDto>? = null,
)

@JsonClass(generateAdapter = true)
data class MailHomeMemberDto(
    val userId: String,
    val name: String? = null,
    val role: String? = null,
)

// POST /api/mailbox/send

@JsonClass(generateAdapter = true)
data class SendMailBody(
    val recipientUserId: String? = null,
    val recipientHomeId: String? = null,
    val type: String = "letter",
    val subject: String,
    val content: String,
    val `object`: SendMailObject,
    val expiresAt: String? = null,
)

@JsonClass(generateAdapter = true)
data class SendMailObject(
    val format: String = "mailjson_v1",
    val title: String? = null,
    val content: String? = null,
    val payload: SendMailPayload,
)

@JsonClass(generateAdapter = true)
data class SendMailPayload(
    val stationeryTheme: String,
    val inkSelection: String,
    val sealChoice: String,
    val intent: String,
    val returnAddressShared: Boolean,
    val voicePostscriptUri: String? = null,
)

@JsonClass(generateAdapter = true)
data class SendMailResponse(
    val message: String? = null,
    val mail: SentMailDto? = null,
)

@JsonClass(generateAdapter = true)
data class SentMailDto(
    val id: String? = null,
    val subject: String? = null,
    @Json(name = "created_at") val createdAt: String? = null,
)

File: frontend/apps/ios/Pantopus/Core/Networking/Models/MailCompose/MailComposeDTOs.swift
//
//  MailComposeDTOs.swift
//  Pantopus
//
//  Decoder shapes for the T3.7 Ceremonial Mail Compose flow.
//

import Foundation

/// Envelope from `GET /api/mailbox/compose/recipients`.
public struct MailComposeRecipientsResponse: Decodable, Sendable {
    public let recipients: [MailRecipientDTO]
}

public struct MailRecipientDTO: Decodable, Sendable, Hashable, Identifiable {
    public let userId: String
    public let name: String?
    public let username: String?
    public let homeId: String?
    public let homeAddress: String?
    public let isVerified: Bool?
    public let homeMediaUrl: String?
    public let isOnPantopus: Bool?

    public var id: String {
        userId
    }
}

/// Envelope from `GET /api/mailbox/compose/home-context/:homeId`.
public struct MailHomeContextResponse: Decodable, Sendable {
    public let homeId: String?
    public let addressDisplay: String?
    public let memberCount: Int?
    public let homeMediaUrl: String?
    public let privateDeliveryAvailable: Bool?
    public let members: [MailHomeMemberDTO]?
}

public struct MailHomeMemberDTO: Decodable, Sendable, Hashable {
    public let userId: String
    public let name: String?
    public let role: String?
}

/// Response from `POST /api/mailbox/send`. We only need the new
/// mail id (in `mail.id`) plus the optional success message.
public struct SendMailResponse: Decodable, Sendable {
    public let message: String?
    public let mail: SentMailDTO?
}

public struct SentMailDTO: Decodable, Sendable, Hashable {
    public let id: String?
    public let subject: String?
    public let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id, subject
        case createdAt = "created_at"
    }
}
